nxpprog.py

```diff
@@ -398,6 +399,81 @@
 prog = None
 
 
+LPC_CHAR = {'Question': b'?',
+                'Synchronized': b'Synchronized\r\n',
+                'OK': b'OK\r\n'}
+
+
+class AutoLPCPortFinder:
+    def __init__(self, device=None):
+        self.device = device
+
+    def port_write_and_verify(self, port, payload, error_message="", debug_message=""):
+        bytes_sent = port.write(bytearray(payload))
+        if bytes_sent != len(payload):
+            log(error_message)
+        else:
+            log(debug_message)
+
+
+    def port_read(self, port, number_of_bytes):
+        return bytearray(port.read(number_of_bytes))
+
+    def find_lpc_port(self):
+        log('Looking for ports with LPC40xx devices')
+        port = None
+        if self.device:
+            serial_device_list = [
+                {
+                    "device": self.device
+                }
+            ]
+        else:
+            serial_device_list = list(port_list.comports())
+
+        for port_info in serial_device_list:
+            try:
+                port = serial.Serial(
+                    port=port_info.device,
+                    baudrate=115200,
+                    parity=serial.PARITY_NONE,
+                    stopbits=serial.STOPBITS_ONE,
+                    bytesize=serial.EIGHTBITS,
+                    timeout=3)
+                sync_successful = self.attempt_sync(port)
+                if sync_successful:
+                    port.close()
+                    return port_info.device
+            except serial.SerialException:
+                log(f'excpetion on port {port}')
+                continue
+
+        return port
+
+    def attempt_sync(self, port):
+        serial_device = SerialDevice(port.port, port.baudrate)
+        serial_device.isp_mode()
+        log(f'Attempting to sync with device at {port.port}')
+        self.port_write_and_verify(port, LPC_CHAR['Question'])
+        time.sleep(.100)
+        response = self.port_read(port, 14)
+
+        if response != LPC_CHAR['Synchronized']:
+            return False
+        time.sleep(.100)
+        self.port_write_and_verify(port, LPC_CHAR['Synchronized'])
+        time.sleep(.100)
+        response = self.port_read(port, 17)
+        if response != LPC_CHAR['Synchronized']:
+            return False
+        self.port_write_and_verify(port, b'115200\r\n')
+        time.sleep(.100)
+        response = self.port_read(port, 4)
+        return response == LPC_CHAR['OK']
+
+
```

Comment:
Remove two of the new lines here. Only should have a distance of 2 newlines between blocks of code.

Reply:
Done.